secure/net/src/grpc_network_service/mod.rs

```diff
@@ -0,0 +1,194 @@
+            .build()
+            .unwrap();
+
+        info!("Starting Server async at {:?}", server_addr);
+        // NOTE: serve_with_shutdown() starts the server, if successful the task does not return
+        // till the server is shutdown. Hence this should be called as a separate non-blocking task.
+        // Signal handler 'server_shutdown_rx' is needed to shutdown the server
+        Server::builder()
+            .add_service(RemoteExecutionServer::new(self))
+            .add_service(reflection_service)
+            .serve_with_shutdown(server_addr, async {
+                server_shutdown_rx.await.ok();
+            })
+            .await
+            .unwrap();
+        info!("Server shutdown at {:?}", server_addr);
+    }
+}
+
+#[tonic::async_trait]
+impl RemoteExecution for RemoteExecutionServerWrapper {
+    async fn simple_msg_exchange(
+        &self,
+        request: Request<NetworkMessage>,
+    ) -> Result<Response<Empty>, Status> {
+        let network_message = request.into_inner();
+        let sender = network_message.sender_addr;
+        let msg = Message::new(network_message.message);
+        let message_type = MessageType::new(network_message.message_type);
+
+        if let Some(handler) = self.inbound_handlers.lock().unwrap().get(&message_type) {
+            // Send the message to the registered handler
+            handler.send(msg).unwrap();
+        } else {
+            error!(
+                "No handler registered for sender: {:?} and msg type {:?}",
+                sender, message_type
+            );
+        }
+        Ok(Response::new(Empty {}))
+    }
+}
+
+pub struct RemoteExecutionClientWrapper {
+    remote_addr: String,
+    remote_channel: RemoteExecutionClient<Channel>,
+}
+
+impl RemoteExecutionClientWrapper {
+    pub fn new(rt: &Runtime, remote_addr: SocketAddr) -> Self {
+        Self {
+            remote_addr: remote_addr.to_string(),
+            remote_channel: rt
+                .block_on(async { Self::get_channel(format!("http://{}", remote_addr)).await }),
+        }
+    }
+
+    async fn get_channel(remote_addr: String) -> RemoteExecutionClient<Channel> {
+        info!("Trying to connect to remote server at {:?}", remote_addr);
+        let conn = tonic::transport::Endpoint::new(remote_addr)
+            .unwrap()
+            .connect_lazy();
+        RemoteExecutionClient::new(conn)
+    }
+
+    pub async fn send_message(
+        &mut self,
+        sender_addr: SocketAddr,
+        message: Message,
+        mt: &MessageType,
+    ) {
+        let request = tonic::Request::new(NetworkMessage {
+            sender_addr: sender_addr.to_string(),
+            message: message.data,
+            message_type: mt.get_type(),
+        });
+        // TODO: Retry with exponential backoff on failure
+        match self.remote_channel.simple_msg_exchange(request).await {
+            Ok(_) => {},
+            Err(e) => {
```

Comment:
Hmm we should just panic in this case, right?

Reply:
Ideally we should have retries. There could be temporary failures on server. We can panic for now.

Reply:
You can add some retry with exponetial back-off I assume..and panic once we exhaust the retries..

Reply:
Adding retries requires an additional copy of msg to send. 
(This can be achieved by existing retry_async() library)

Since we do not want to bite the bullet of the cost of additional copy, leaving retry as a TODO for now